### DISCUSSION TITLE
Share your Suggestions, Feedback & Improvements for API here.

### DISCUSSION THREAD

**Discussion by dr5hn** [dr5hn/countries-states-cities-database#151] [Ideas]
Hello all 👋🏼 ,

I would like to request everyone who is using this [API](https://countrystatecity.in/), please share their suggestions, feedbacks & improvements.

**Comment by samihoda** [comment]
How do I request an API key? The site says no more submissions.

**Reply by dr5hn** [reply]
@samihoda My Free TypeForm Limits went out today,

So Share your email here, I will send it across an API Key.

P.S.: Typeform limits reset every 8th of the month.

**Comment by samihoda** [comment]
@dr5hn I will email you directly in a minute. Appreciate it!

**Reply by dr5hn** [reply]
@samihoda 
I have sent you API Key as a reply to your email.

**Comment by samihoda** [comment]
Thanks. I noticed the docs do not list the endpoint. I can assume its api.... but I couldn't find the actual endpoint.

**Reply by dr5hn** [reply]
@samihoda Let me know if these links solves your query.

https://countrystatecity.in/docs/api/all-countries
https://countrystatecity.in/docs/api/all-states
https://countrystatecity.in/docs/api/states-by-country
https://countrystatecity.in/docs/api/state
https://countrystatecity.in/docs/api/cities-by-state-country
https://countrystatecity.in/docs/api/cities-by-country

**Comment by samihoda** [comment]
@dr5hn No I mean the API endpoint, meaning what URL the API call should hit. I am assuming its https://api.count.... but it is not documented anywhere, just the path is /v1...

**Reply by dr5hn** [reply]
@samihoda Yeah may be i missed it.. my bad :(

The base url for api is
```
https://api.countrystatecity.in
```

**Reply by dr5hn** [reply]
@samihoda Thanks for your findings. I have fixed it.

https://countrystatecity.in/docs/api/all-countries/

**Comment by samihoda** [comment]
Thanks, this looks good. The Swagger playground seems to work, however I can't get it to due to CORS. Trying to figure out the issue.

**Reply by dr5hn** [reply]
@samihoda CORS Issue Fixed.

**Comment by samihoda** [comment]
Access to XMLHttpRequest at 'https://api.countrystatecity.in/v1/countries/IN/states' from origin '...' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: It does not have HTTP ok status.

**Reply by dr5hn** [reply]
@samihoda can you please send me your implementation files so that i can check ?

**Reply by samihoda** [reply]
I sent it via email. Thanks!

**Reply by dr5hn** [reply]
@samihoda CORS Issue Fixed.

**Reply by samihoda** [reply]
@dr5hn It works! I'll make changes in my site and give it a good QA round, will let you know if I encounter any additional issues. Thanks.

**Comment by aakashmartand** [comment]
Hello Darshan,
It's nice to have this database, it's very helpful.
But there are some repetitions in the city names, for eg. Bikaner, Hanumangarh, Pratapgarh, etc.
Kindly look into that.

**Reply by dr5hn** [reply]
@aakashmartand Thanks for your findings.

I will look into it asap.

**Reply by dr5hn** [reply]
@aakashmartand It's fixed now. 👍🏼

**Comment by renegng** [comment]
Hi @dr5hn amazing API! Love it! Just one thing I noticed: For Honduras Departments, "Bay Island" should be named "Isla de la Bahía" so that it has consistency with the rest of the Departments (since the rest are in spanish too). Thank you so much for your work!

Maybe implementing an option for Database language would be an interesting next version. Let me know if I could help ok!

Best regards!
René

**Reply by dr5hn** [reply]
@renegng Thanks for suggestions. It's already on my list. 😄 


I will fix this ASAP 👉🏼 "Bay Island" should be named "Isla de la Bahía"

**Comment by aputra20** [comment]
Hi, would like to try it, can help to create the API key?

Thanks !

**Reply by dr5hn** [reply]
@aputra20 Check your emails 🎉

**Comment by abhinavsindhu1997** [comment]
abhinavsindhu1997@gmail.com
Please send the API Key.

**Reply by dr5hn** [reply]
Sent !! 🎉

**Comment by Haaziq-Uvais** [comment]
Hi @dr5hn, thank you for this amazing API. I have one request, can you send the longitude and latitude of the cities in the Cities by country API? It's there in the JSON but the API only returns the id and the name of the city. Thanks.

**Reply by dr5hn** [reply]
Hi @Haaziq-Uvais,

I have purposely not added those data, because due to lots of traffic, current server configurations are not capable of handling this additional data.

Also, I don't get enough donations to make this server maintain & upgrade.

So I am going to add these functionalities in the Paid version. However, I am going to keep this FREE version as it is forever (with data updates but no feature additions).

Hope that makes sense!

**Reply by Haaziq-Uvais** [reply]
Yes, that makes perfect sense! Is the paid API currently available or if not, any idea when will it be available?

**Reply by dr5hn** [reply]
I am still working on it! No ETA at the moment.

**Comment by akshay1515** [comment]
Hey Darshan can you help me with modified json as i require it to build my prototype app to some investors.
i want json of cities by country api response as 
{
id: cityid,
name: cityname,
state: statename,
country: countryname,
zipcode : cityzipcode
}

**Reply by dr5hn** [reply]
@akshay1515 If you know PHP you can get it done by yourself using this script.
https://github.com/dr5hn/countries-states-cities-database/blob/master/scripts/export_csc_npm.php

Cheers.

**Comment by XDavidT** [comment]
I want to suggest improvement!
Most of the people who build forms and need this API will also need a country phone prefix, maybe add optional request to get phone code with the request together?

**Comment by faisalcodes** [comment]
It would be very helpful if you add full text search feature in the api. If one searches for a particular term then the related cities, states or countries in the selected category should be returned in the response.

**Reply by dr5hn** [reply]
@faisalcodes Thanks for the suggestion.

Did you mean open search? Kind of searching in all columns of the data?

If yes, Sure will add it to the PRO version. As this will lead to performative operations this can't be provided in the FREE version.

No ETA for PRO.

**Comment by CJ-0221** [comment]
Hi @dr5hn ,

Amazing API and thanks for sharing APIKEY very fast. I have a small question the api is not giving all the information on countries list for eg: Flag, iso3 etc. It giving very limited info. do we have to include any to get all the details. please suggest.

**Reply by dr5hn** [reply]
@CJ-0221 You can get that using CountryDetail API
https://countrystatecity.in/docs/api/country/

I have kept all countries API to minimal information for performance purpose.

**Comment by spl-laswi** [comment]
Hi @dr5hn,

Thanks for your hard and fantastic work on this API, I have a question: Is there a way to map the city with the state?
For example:
states = v1/countries/US/states;
cities = v1/countries/US/cities;
Then map the cities inside each state by id.
I don't want to send a request for each state to get the cities of it.

**Comment by bensonarafat** [comment]
The API is cool, but it will be nice to have relationships
Like when fetch country, to have the option to fetch city along

**Reply by dr5hn** [reply]
There's an API route already available where you can supply country code and get the cities of that country.

**Reply by bensonarafat** [reply]
Yeah, i know of that.. 
But what i trying to do, is fetch all the countries with their cities too. with just one request. 

The way it is now, i have to fetch the country then fetch the cities.. that's double request i have to send.

**Reply by dr5hn** [reply]
@bensonarafat I would advise against doing that as it could increase the weight of the response. A heavier response could potentially impact the response time and burden the server load.

**Comment by sk7683** [comment]
In the cities by country API, Is it possible to add the state code (iso2) and id (state primary key) in the response object? 

https://api.countrystatecity.in/v1/countries/IN/cities

**Comment by JavaMule** [comment]
The APIs are great, I really appreciate your work. Is there any Database table or API for ZIP Codes/Pincodes as below?
Table/API with zipcode/pincode, city, state, country.

**Comment by lillyYu** [comment]
Hello I found this API super easy to use and helpful for building multi-language site. I have one request on this api.

How about adding translations for states and cities like you do have for the countries. I want to show translated label for each user who visit my site. 

Thank you for your awesome work!

**Comment by mcl123** [comment]
https://countrystatecity.in/request   API key why not found?
Thank you!

**Reply by dr5hn** [reply]
@mcl123 try https://forms.gle/ckZ8Gub6jS9sUihJA

**Comment by ppfoong** [comment]
I suggest to add in a new boolean column in cities table called `is_capital` and use it to tag whether the city is a capital of its country.

After that, you can update the field in cities table by using this SQL command:

UPDATE `cities`
  SET `is_capital` = 1
  WHERE EXISTS (SELECT * FROM `countries` WHERE `countries`.`capital` = `cities`.`name` AND `countries`.`id` = `cities`.`country_id`);

and then, you can also add a new `capital_id` column in countries table, and store the capital city's id in that field.

**Comment by abhishek-kumar-96** [comment]
Hey, these are really great APIs

Have one request for Country cities API, that 

1. it should support pagination, a country could have thousands of cities mayb not be good idea to load them all
2. If possible maybe a newer API with city search functionality, like provide Country ISO and a query and it gives out matching (fuzzy) cities.

Thanks and keep up the great work.

**Comment by CeiyoneAshok** [comment]
I need the latitude and longitude of the cities in the response

**Comment by vedicastrology123** [comment]
I had raised a request for API key through google forms.
Yet to receive.

Please help.

**Comment by eyahellal** [comment]
thank you so much your api has been very helpful

**Comment by uptk-umang** [comment]
I’m using your [Country-State-City API](https://countrystatecity.in/docs/) and it’s been very helpful.
I have a feature request that I believe could improve the API for international applications:

Would it be possible to provide translation data in the list endpoints for countries, states, and cities? Ideally, this could be implemented in one of the following ways:

1. Allow passing a query parameter like **locale** or **language** to return names in the requested language.
2. Include an optional translations field in the response that contains name translations for supported languages.

This would greatly simplify the process for apps that need to display localized names without having to maintain a separate mapping.
Thank you for considering this feature. I’d love to hear your thoughts or if there are any current plans to support this.

**Comment by yeahfine** [comment]
Sorry, I would like to know what criteria you used to precisely divide 250 countries, 5,038 states, and 151,024 cities.

**Comment by yeahfine** [comment]
What I must tell you is that there is a serious error in your administrative division: **Taiwan is a part of China and not an independent country!!!** **Similarly, the Hong Kong Special Administrative Region and the Macau Special Administrative Region are also part of China, and you cannot directly classify them as countries!!!** Please correct this error immediately!!!
